### PR TITLE
[Yinoussa] Modification to make this compatible with fix constraint in SoftRobot.Inverse pluging.

### PR DESCRIPTION
**Pull request by Younesssss**

### CODE REVIEW

File: src/SoftRobots/component/constraint/CableConstraint.cpp [hunk 1]
```diff
@@ -113,6 +113,8 @@ void CableForceConstraintResolution::resolution(int line, double** w, double* d,
     }
     else
         lambda[line] = m_imposedForce;
+
+    std::cout << "The force at the END is : " << lambda[line] << std::endl;
```
Comment (EulalieCoevoet) [inline src/SoftRobots/component/constraint/CableConstraint.cpp:117]: To remove I guess

File: src/SoftRobots/component/constraint/SurfacePressureConstraint.inl [hunk 2]
```diff
@@ -183,7 +183,7 @@ void SurfacePressureConstraint<DataTypes>::setUpPressureLimits(double& imposedVa
 {
     if (d_maxPressureVariation.isSet())
     {
-        double pressure = d_pressure.getValue();
+        double pressure = d_maxPressure.getValue();
```
Comment (EulalieCoevoet) [inline src/SoftRobots/component/constraint/SurfacePressureConstraint.inl:186]: I feel like this is wrong

File: src/SoftRobots/component/constraint/model/CableModel.inl [hunk 3]
```diff
@@ -90,13 +90,22 @@ CableModel<DataTypes>::CableModel(MechanicalState* object)
                           "If unspecified no minimum value will be considered \n"
                           "and the cable will then be seen as a stiff rod able to push."))
 
+    , d_eqForce(initData(&d_eqForce,Real(0.0), "eqForce",
+                          "Equality force of the actuator. \n"
+                          "Solver will try to maintain the cable force at this value\n"
+                          "If unspecified value will be considered \n"
```
Comment (EulalieCoevoet) [inline src/SoftRobots/component/constraint/model/CableModel.inl:96]: typo? "If unspecified, no value will be considered."?

File: src/SoftRobots/component/constraint/model/CableModel.inl [hunk 4]
```diff
@@ -90,13 +90,22 @@ CableModel<DataTypes>::CableModel(MechanicalState* object)
...
     , d_maxPositiveDisplacement(initData(&d_maxPositiveDisplacement,Real(0.0), "maxPositiveDisp",
                                          "Maximum displacement of the actuator in the positive direction. \n"
                                          "If unspecified no maximum value will be considered."))
 
     , d_maxNegativeDisplacement(initData(&d_maxNegativeDisplacement,Real(0.0), "maxNegativeDisp",
                                          "Maximum displacement of the actuator in the negative direction. \n"
                                          "If unspecified no maximum value will be considered."))
+    , d_eqDisplacement(initData(&d_eqDisplacement,Real(0.0), "eqDisp",
+                                         "Specify cable displacement (equality) . \n"
+                                         "If unspecified no value will be considered."))
```
Comment (EulalieCoevoet) [inline src/SoftRobots/component/constraint/model/CableModel.inl:108]: I suggest a description similar to d_eqForce, which is really nice: 
```
"Equality displacement of the actuator. \n"  
"Solver will try to maintain the cable displacement at this value\n"  
"If unspecified, no value will be considered \n"  
```

File: src/SoftRobots/component/constraint/model/QPSlidingConstraint.h [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+    double      m_wActuatorActuator;
+    double      m_imposedForce;
+    double      m_minDisplacement;
+    double      m_maxDisplacement;
+
+};
+
+
+/**
+ * This class contains common implementation of cable constraints
```
Comment (EulalieCoevoet) [inline src/SoftRobots/component/constraint/model/QPSlidingConstraint.h:82]: to update?

File: src/SoftRobots/component/constraint/model/QPSlidingConstraint.h [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+*/
+template< class DataTypes >
+class QPSlidingConstraint : public CableModel<DataTypes>
```
Comment (EulalieCoevoet) [inline src/SoftRobots/component/constraint/model/QPSlidingConstraint.h:85]: Why QP in the name? Couldn't it be just "SlidingConstraint"?

File: src/SoftRobots/component/constraint/model/QPSlidingConstraint.inl [hunk 7]
```diff
@@ -0,0 +1,292 @@
...
+template<class DataTypes>
+void QPSlidingConstraint<DataTypes>::init()
+{
+    Inherit1::init();
+
+    // To remove in SoftRobots v20.0
+    if(!d_minForce.isSet())
+        msg_warning() << "An old implementation of CableConstraint was not allowing negative force. This is now possible. "
+                      << "However, to limit the force to be strictly positive you now have to set minForce=0.";
+    //
```
Comment (EulalieCoevoet) [inline src/SoftRobots/component/constraint/model/QPSlidingConstraint.inl:95]: I guess we can remove this now :)

File: src/SoftRobots/component/constraint/model/QPSlidingConstraint.inl [hunk 8]
```diff
@@ -0,0 +1,292 @@
...
+            MatrixDerivRowIterator c_it = matrix.writeLine(cIndex);
+            c_it.addCol(i, Coord(1,0,0));
+
+            MatrixDerivRowIterator c_it_1 = matrix.writeLine(cIndex+1);
+            c_it_1.addCol(i, Coord(0,1,0));
+            MatrixDerivRowIterator c_it_2 = matrix.writeLine(cIndex+2);
+            c_it_2.addCol(i, Coord(0,0,1));
+            cIndex +=3;
+        }
+
```
Comment (EulalieCoevoet) [inline src/SoftRobots/component/constraint/model/QPSlidingConstraint.inl:155]: What does this component do exactly? I'm not sure to understand.

File: src/SoftRobots/component/constraint/model/QPSlidingConstraint.inl [hunk 9]
```diff
@@ -0,0 +1,292 @@
...
+        //std::cout << "Size JDX = "<< Jdx->size() << std::endl;
+        for (size_t i = 0; i < positions.size(); i++){
+            if( i < positions.size()-1){
+                Real dfree1 = Jdx->element(2*i)   + positions[i][1];
+                Real dfree2 = Jdx->element(2*i+1) + positions[i][2];
+
+                resV->set(m_constraintId + 2*i   , dfree1);
+                resV->set(m_constraintId + 2*i +1, dfree2);
+            }else{
+                //std::cout << " The laste position : "<< positions[i] << "; Jdx->element() "<< Jdx->element(2*i) <<" "<< Jdx->element(2*i+1) <<" "<< Jdx->element(2*i+2) << std::endl;
```
Comment (EulalieCoevoet) [inline src/SoftRobots/component/constraint/model/QPSlidingConstraint.inl:203]: remove the commented debug lines?

File: src/SoftRobots/component/constraint/model/SurfacePressureModel.inl [hunk 10]
```diff
@@ -92,11 +92,11 @@ SurfacePressureModel<DataTypes>::SurfacePressureModel(MechanicalState* object)
     , d_minPressure(initData(&d_minPressure, Real(0.0), "minPressure",
                              "Minimum pressure allowed for actuation. If no value is set by user, no \n"
                              "minimum pressure constraint will be considered. A negative pressure will empty/drain the cavity."))
-
+    , d_eqPressure(initData(&d_eqPressure, Real(0.0), "eqPressure",
+                             "Pressure is set to a value desired by the user (equality)\n"))
```
Comment (EulalieCoevoet) [inline src/SoftRobots/component/constraint/model/SurfacePressureModel.inl:96]: Same here, something like
```
"Equality constraint for the pressure. \n"  
"Solver will try to maintain the pressure at this value\n"  
"If unspecified, no equality constraint will be considered \n"  
```

File: src/SoftRobots/component/constraint/model/SurfacePressureModel.inl [hunk 11]
```diff
@@ -115,6 +115,9 @@ SurfacePressureModel<DataTypes>::SurfacePressureModel(MechanicalState* object)
                              "maximum will be considered. NB: this value has a dependancy with the time step \n"
                              "(volume/dt) in the dynamic case."))
 
+    , d_eqVolumeGrowth(initData(&d_eqVolumeGrowth, Real(0.0), "eqVolumeGrowth",
+                             "Volume is set to a value desired by the user (equality) \n"))
+
```
Comment (EulalieCoevoet) [inline src/SoftRobots/component/constraint/model/SurfacePressureModel.inl:120]: same here

**Comment by EulalieCoevoet** [comment]
What does the component `QPSlidingConstraint` do exactly? I'm not sure to understand.

File: CMakeLists.txt [hunk 12]
```diff
@@ -56,7 +56,10 @@ set(HEADER_FILES
     ${SOFTROBOTS_SOURCE_DIR}/component/behavior/SoftRobotsBaseConstraint.h
     ${SOFTROBOTS_SOURCE_DIR}/component/behavior/SoftRobotsConstraint.h
     ${SOFTROBOTS_SOURCE_DIR}/component/behavior/SoftRobotsConstraint.inl
+    ${SOFTROBOTS_SOURCE_DIR}/component/behavior/CosseratSoftRobotsConstraint.h
+    ${SOFTROBOTS_SOURCE_DIR}/component/behavior/CosseratSoftRobotsConstraint.inl
```
Comment (EulalieCoevoet) [inline CMakeLists.txt:60]: These files are not in the PR right?

**Comment by Younesssss** [comment]
@EulalieCoevoet can you have a look? thanks

**Comment by EulalieCoevoet** [comment]
Looks good to me 👍 

Can you think of an easy way to revert your changes to:  
`docs/tutorials/Tripod/details/step8/step8.pyscn`  
`docs/tutorials/Tripod/details/step8/tripod.py `  
`docs/tutorials/Trunk/trunk.py`  

It seems that they just introduce a lot of weird line breaks. Like this:

```
self.node.createObject(
       'MeshVTKLoader', name='loader', filename=path+'trunk.vtk')
```

**Review by EulalieCoevoet — APPROVED** [review]

**Comment by Younesssss** [comment]
For line breaks sorry, that the problem with visual code editor. 
Thanks for fixing it.

**Pull request merged**